tidy3d/plugins/design/design.py

```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+        return self.method.run(run_fn=evaluate_fn, parameters=self.parameters)
+
+    def run_pre_post(self, fn_pre: Callable, fn_post: Callable):
+        """Run a function with tidy3d implicitly called in between."""
+        evaluate_fn = self._get_evaluate_fn_pre_post(fn_pre=fn_pre, fn_post=fn_post)
+        return self.method.run(run_fn=evaluate_fn, parameters=self.parameters)
+
+    """ Helpers """
+
+    def _get_evaluate_fn_single(self, fn: Callable) -> list[Any]:
+        """Get function that sequentially evaluates single `fn` for a list of arguments."""
+
+        def evaluate(args_list: list) -> list[Any]:
+            """Evaluate a list of arguments passed to ``fn``."""
+            return [fn(**args) for args in args_list]
+
+        return evaluate
+
+    def _get_evaluate_fn_pre_post(self, fn_pre: Callable, fn_post: Callable) -> list[Any]:
+        """Get function that tries to use batch processing on a set of arguments."""
+
+        def evaluate(args_list: list[tuple[Any, ...]]) -> list[Any]:
+            """Put together into one pipeline."""
+            combined_fn = self._stitch_pre_post(fn_pre=fn_pre, fn_post=fn_post)
+            out = combined_fn(args_list)
+            return out
+
+        return evaluate
+
+    def _stitch_pre_post(self, fn_pre: Callable, fn_post: Callable) -> Callable:
+        """Combine pre and post into one big function, stitching tidy3d calls between if needed."""
+
+        def fn_combined(args_list):
+            sim_dict = {idx: fn_pre(**arg_list) for idx, arg_list in enumerate(args_list)}
+            data = self.fn_mid(sim_dict)
+            post_out = [fn_post(val[1]) for val in data.items()]
+            return post_out
+
+        return fn_combined
+
+    def fn_mid(self, pre_out: Any) -> Any:
+        """A function of the output of ``fn_pre`` that gives the input to ``fn_post``."""
+
+        # Convert list of sim inputs to dict of dict before passing through checks
+        was_list = False
+        if all(isinstance(sim, List) for sim in pre_out.values()) and any(
+            isinstance(sim, Simulation) for sim in pre_out[0]
+        ):
+            pre_out = {
+                idx: dict(enumerate(sim_list)) for idx, sim_list in enumerate(pre_out.values())
+            }
+            was_list = True
+
+        # Handle most common case where fn_combined builds a dict of sims
+        if all(isinstance(sim, Simulation) for sim in pre_out.values()):
+            named_sims = {next(self._name_generator): sim for sim in pre_out.values()}
+            self._task_names.extend(list(named_sims.keys()))
+            data = web.Batch(simulations=named_sims, folder_name=self.folder_name).run(
+                path_dir=self.path_dir
+            )
+
+        # Can "index" dict here because fn_combined uses idx as the key
+        elif all(isinstance(sim, Dict) for sim in pre_out.values()) and any(
+            isinstance(sim, Simulation) for sim in pre_out[0].values()
```

Comment:
same comment here about only checking first value for sim-ness. we should make sure to handle all three edge cases 1. all sims, 2. no sims, 3. mixed. even if it means erroring for case 3.